### PR TITLE
EP + Trainer integration on top of DistributedConfig (#45028)

### PR DESCRIPTION
**Pull request by AmineDiro**
Builds on [#45028](https://github.com/huggingface/transformers/pull/45028) (`refactor-tp-dtensor`) so that **Expert Parallelism actually runs** on any MoE model that has a `base_ep_plan`:
  ```python
  DistributedConfig(enable_expert_parallel=True, tp_size=N, fsdp_size=M)
  ```

I was validated end-to-end with  `trl.SFTTrainer` on `Qwen3-30B-A3B` at multiple shapes (16k / 32k / 64k, 2n / 4n / 8n).  Healthy training (loss bit-exact match against our prod fork at LR=0) and very competitive MFU 👏🏼 👏🏼 

> **Best run: 64k ctx length,  2 nodes, EP=16 + sonicmoe + Liger + compile = 62.8 % window MFU / 35 % causal-adjusted, peak GPU mem 45 GB**. 

# What does this PR do?
- **Register `ep_router` + `moe_experts_ep_allreduce` ParallelStyles** (`distributed/tensor_parallel.py`). `RouterParallel` masks non-local expert scores.  `MoEExpertsParallel` shards on the experts dim. 
- `model._tp_plan` → `model.tp_plan` so `enable_expert_parallel=True` actually selects `_ep_plan` (the property override at `modeling_utils.py:1459`).
- **`Trainer._clip_grad_norm` routes to the mesh-aware streaming clip** (`trainer.py`) when on the native path. Stock `accelerator.clip_grad_norm_` across cross-mesh DTensor grads crashes on the first `optimizer.step`.
- **`Trainer.create_optimizer` calls `maybe_disable_foreach_and_fused_for_mixed_dtensor_groups`** helper; Adam fused/foreach rejects mixed Tensor/DTensor groups under `fsdp_size=1`.
- **`Trainer.compute_loss` divides `loss_scale` by `tp_size` for the `DistributedConfig` path** (`trainer.py`). Companion to #45994 (which only handles `parallelism_config.tp_size`).
- **Biggest change is that I introduce `Trainer.is_native_distributed` flag**, exclusive with `is_fsdp_enabled` / `is_deepspeed_enabled` @3outeille  tell me what you think.   This is made to make clear when we are running in native vs accelerate accross branches.
- **Qwen3MoE `base_model_ep_plan` simplified** to use `moe_experts_ep_allreduce`: @ArthurZucker  I wonder if we need to have this ep plan dynamically set by the user and not part of the model definition 🤔 

# Note:
This PR depends on : 
- ⚠️ **[#46113 (clip_grad_norm streaming)](https://github.com/huggingface/transformers/pull/46113)** — required. This PR's `_clip_grad_norm` route depends on the streaming O(1)-per-param clip from #46113. Without it, every 30B+ EP+FSDP `optimizer.step` allocates a ~60 GB transient (current `clip_grad_norm` replicates every DTensor grad). So we should land #46113 first.
- **[#45994 (loss-scale fix under TP)](https://github.com/huggingface/transformers/pull/45994)**  PR. This PR extends #45994 to the native FSDP path.

## Who can review?

@3outeille @ArthurZucker @SunMarc

### CODE REVIEW

File: src/transformers/distributed/tensor_parallel.py [hunk 1]
```diff
@@ -451,6 +501,18 @@ class ParallelInterface(GeneralInterface):
...
             ),
+            # MoE — EP variant: shards on the experts dim(0) so each rank holds
+            # `num_experts/ep_size` experts at full hidden_in.
+            "moe_experts_ep_allreduce": MoEExpertsParallel(
+                output_layouts=Replicate(),
+                shard_plan={
+                    "gate_up_proj": Shard(0),
+                    "down_proj": Shard(0),
+                },
+            ),
```
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py:512]: No no no no nonoooooo this is the worst pattern that was ever introduced in the lib, sorry you have to "pay the price" but EP should be back to how we did it: `gate_up_proj` gets "gemm" and `down_proj` as well, te module gets `ep_allreduce`!

File: src/transformers/distributed/utils.py [hunk 2]
```diff
@@ -150,19 +150,37 @@ def distribute_model(model, distributed_config: DistributedConfig, device_mesh)
...
-    FSDP-only params on the (fsdp,) sub-mesh). We sidestep it by replicating each
-    DTensor grad to a plain local tensor, computing the norm over those, and
-    scaling the original DTensor grads in place — the placement of the original
-    grads doesn't matter for the per-element clip.
-    """
-    grads = [p.grad for p in parameters if p.grad is not None]
-    local_grads = [_replicate_dtensor(g).to_local() if isinstance(g, DTensor) else g for g in grads]
-    total_norm = torch.nn.utils.get_total_norm(local_grads, norm_type=norm_type)
-    torch.nn.utils.clip_grads_with_norm_(grads, max_norm=max_norm, total_norm=total_norm)
+    """Mesh-aware grad-norm clip with O(1) extra memory per parameter."""
```
Comment (ArthurZucker) [inline src/transformers/distributed/utils.py:153]: remove unrelated if unrelated

File: src/transformers/models/qwen3_moe/configuration_qwen3_moe.py [hunk 3]
```diff
@@ -86,9 +86,14 @@ class Qwen3MoeConfig(PreTrainedConfig):
     # This allows EP to scale beyond num_kv_heads (not constrained by 4 for Qwen3-30B).
     base_model_ep_plan = {
         "layers.*.mlp.gate": "ep_router",
-        "layers.*.mlp.experts.gate_up_proj": "grouped_gemm",
-        "layers.*.mlp.experts.down_proj": "grouped_gemm",
-        "layers.*.mlp.experts": "moe_experts_allreduce",
```
Comment (ArthurZucker) [inline src/transformers/models/qwen3_moe/configuration_qwen3_moe.py:91]: this is what we want

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -816,9 +818,25 @@ def create_accelerator_and_postprocess(self) -> None:
...
+        _ds = getattr(self.accelerator.state, "deepspeed_plugin", None) is not None
+        _fsdp = getattr(self.accelerator.state, "fsdp_plugin", None) is not None
+        if self.is_native_distributed and (_ds or _fsdp):
+            logger.warning_once(
+                "`distributed_config` is set on the model AND accelerate has a "
+                f"{'deepspeed' if _ds else 'fsdp'}_plugin in state. Trusting `distributed_config` "
+                "(transformers-native) and ignoring the accelerate plugin."
+            )
+        self.is_deepspeed_enabled = _ds and not self.is_native_distributed
+        self.is_fsdp_enabled = _fsdp and not self.is_native_distributed
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:839]: not super clean IMO but in trainer soooo

**Review by ArthurZucker — COMMENTED** [review]
Ty, let's not push for wrong patterns

File: src/transformers/distributed/tensor_parallel.py (hunk 1 above)
Comment (AmineDiro) [reply src/transformers/distributed/tensor_parallel.py:512]: no worries 🤣 ! Ok, I was just following the exact implementation pattern: https://github.com/huggingface/transformers/blob/a65bf6c9d03322c1bf1963ef7c64289c4b3f0757/src/transformers/distributed/tensor_parallel.py#L447

Just sharding on the dim 0 for the experts. If I go back to : 
```python
        "layers.*.mlp.experts.gate_up_proj": "grouped_gemm",
        "layers.*.mlp.experts.down_proj": "grouped_gemm",
        "layers.*.mlp.experts": "moe_experts_allreduce",
```

There is some implementation logic that uses the new `TensorParallelStyle` that needs to be rewritten, I guess

File: src/transformers/distributed/utils.py (hunk 2 above)
Comment (AmineDiro) [reply src/transformers/distributed/utils.py:153]: Oh yes I created I different PR : https://github.com/huggingface/transformers/pull/46113 for the grad clipping issue

File: src/transformers/trainer.py (hunk 4 above)
Comment (AmineDiro) [reply src/transformers/trainer.py:839]: Hah, agreed on cleanliness, but let me defend this 😆. I just tried to localize variables that clearly indicate which mode we are running, so we can do stuff like `self.is_deepspeed_enabled or self.is_fsdp_enabled or self.is_native_distributed`. Because we can be running in fsdp from accelerate of Zero3 from native and maybe have passed the `DistributedConfig,` we just want to avoid overlap and be explicit in the condition. 
But happy to change this if you have a better idea!
